streaming/database.js

```diff
@@ -1,107 +1,91 @@
-    }
-    if (typeof parsedUrl.database === 'string') baseConfig.database = parsedUrl.database;
-    if (typeof parsedUrl.options === 'string') baseConfig.options = parsedUrl.options;
-
-    // The pg-connection-string type definition isn't correct, as parsedUrl.ssl
-    // can absolutely be an Object, this is to work around these incorrect
-    // types, including the casting of parsedUrl.ssl to Record<string, any>
-    if (typeof parsedUrl.ssl === 'boolean') {
-      baseConfig.ssl = parsedUrl.ssl;
-    } else if (typeof parsedUrl.ssl === 'object' && !Array.isArray(parsedUrl.ssl) && parsedUrl.ssl !== null) {
-      /** @type {Record<string, any>} */
-      const sslOptions = parsedUrl.ssl;
-      baseConfig.ssl = {};
-
-      baseConfig.ssl.cert = sslOptions.cert;
-      baseConfig.ssl.key = sslOptions.key;
-      baseConfig.ssl.ca = sslOptions.ca;
-      baseConfig.ssl.rejectUnauthorized = sslOptions.rejectUnauthorized;
+    // parse will throw if both useLibpqCompat option is true and the
+    // DATABASE_URL includes uselibpqcompat, so we're handling that case ahead
+    // of time to give a more specific error message:
+    if (env.DATABASE_URL.includes('uselibpqcompat')) {
+      throw new Error(
+        'SECURITY WARNING: Mastodon forces uselibpqcompat mode, do not include it in DATABASE_URL',
+      );
     }
 
-    // Support overriding the database password in the connection URL
-    if (!baseConfig.password && env.DB_PASS) {
-      baseConfig.password = env.DB_PASS;
-    }
+    config = toClientConfig(parse(env.DATABASE_URL, { useLibpqCompat: true }));
   } else if (Object.hasOwn(pgConfigs, environment)) {
-    baseConfig = pgConfigs[environment];
+    config = pgConfigs[environment];
 
     if (env.DB_SSLMODE) {
-      switch(env.DB_SSLMODE) {
-      case 'disable':
-      case '':
-        baseConfig.ssl = false;
-        break;
-      case 'no-verify':
-        baseConfig.ssl = { rejectUnauthorized: false };
-        break;
-      default:
-        baseConfig.ssl = {};
-        break;
+      logger.warn(
+        'Using DB_SSLMODE is not recommended, instead use DATABASE_URL with SSL options',
+      );
+
+      switch (env.DB_SSLMODE) {
+        case 'disable': {
+          config.ssl = false;
+          break;
+        }
+        case 'prefer': {
+          config.ssl.rejectUnauthorized = false;
+          break;
+        }
+        case 'require': {
+          config.ssl.rejectUnauthorized = false;
+          break;
+        }
+        case 'verify-ca': {
+          throw new Error(
+            'SECURITY WARNING: Using sslmode=verify-ca requires specifying a CA with sslrootcert. If a public CA is used, verify-ca allows connections to a server that somebody else may have registered with the CA, making you vulnerable to Man-in-the-Middle attacks. Either specify a custom CA certificate with sslrootcert parameter or use sslmode=verify-full for proper security. This can only be configured using DATABASE_URL.',
+          );
```

Comment:
I see how this is security-related, but not how it is a warning.

Reply:
Could error out? Basically DB_SSLMODE needs additional inputs to be used securely